### PR TITLE
implement fscrypt for PAM

### PR DESCRIPTION
**Pull request by NeoTheFox**
Adresses  #92352

Tested on a running NixOS 20.90

### CODE REVIEW

File: nixos/modules/security/pam.nix
Comment (andir) [inline nixos/modules/security/pam.nix:875]: can you properly indent this?

**Review by andir — CHANGES_REQUESTED** [review]
Thanks for the contribution. Only had a brief look at this would like you to take a moment to reformat the changes in this PR.

You seem to be mixing tabs and spaces whereas we are (mostly) only using spaces in this repository. Usually these settings can be automatically applied by your editor (so you can keep on pressing TAB but still emit the correct number of spaces) by telling your to read `editorconfig` files. There are a lot of plugins available for most of the common editors: https://editorconfig.org/#download

Furthermore you can probably squash all of the changes into one commit that follows our [commit style guide](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes).
In your case this is likely something like this:
```
nixos/security/pam: support fscrypt

This PR adds support for doing XYZ using fscrypt in combination with PAM. This is desireable because reasons ….
```

File: nixos/modules/security/pam.nix
Comment (NeoTheFox) [reply nixos/modules/security/pam.nix]: I removed all tabs

**Review by kamentomov — COMMENTED** [review]
@NeoTheFox I would like to add some changes to this Pull Request (already tested). Alternateively I can comment here about them. If you prefer the former could you please grant me access to your repo? You need to check the box in the pull request that says “Allow edits from maintainers.”

**Comment by NeoTheFox** [comment]
The box was checked! I'm ok with any changes as long as fscrypt gets implemented

**Comment by kamentomov** [comment]
> The box was checked! I'm ok with any changes as long as fscrypt gets implemented

I continue not to have rights. Could you please allow access to contributors in your fork so that I try to commit in NeoTheFox:master?

**Comment by primeos** [comment]
> I continue not to have rights.

@kamentomov “Allow edits from maintainers.” won't work for you because you need commit access to Nixpkgs. I'd probably be better to suggest changes via the GH UI or link to your commit / post a patch so that @NeoTheFox could cherry-pick/apply it.

But in any case I'm not sure if this PR can get merged anytime soon (I'd recommend to find/ping some more appropriate reviewers).
The thing's I'm a bit concerned about:
- `fscrypt-experimental` is named that way because there isn't a stable release yet: https://github.com/google/fscrypt#note-about-stability
- There are different options to enable the PAM module and they have different trade-offs: https://github.com/google/fscrypt#enabling-the-pam-module-on-other-linux-distros
  - I.e. we'd need to decide what's the best default for Nixpkgs.
- I'm not sure if this belongs to `nixos/modules/security/pam.nix` (there's already a comment that "Most of these should be moved to specific modules.")

**Comment by kamentomov** [comment]
@primeos Thanks for commenting. This looked so dead.

> @kamentomov “Allow edits from maintainers.” won't work for you because you need commit access to Nixpkgs. I'd probably be better to suggest changes via the GH UI or link to your commit / post a patch so that @NeoTheFox could cherry-pick/apply it.

I pushed to @NeoTheFox's repo and that worked. @NeoTheFox  thanks for graning me access.

> But in any case I'm not sure if this PR can get merged anytime soon (I'd recommend to find/ping some more appropriate reviewers).

Could you please help us with this? I don't know who is qualified to review this.

> The thing's I'm a bit concerned about:
> 
> * `fscrypt-experimental` is named that way because there isn't a stable release yet: https://github.com/google/fscrypt#note-about-stability

Stragely the package is not experimental in other OSes. Nevertheless, `experimental` in the name seems good enough for me. I mean I don't think we shall do something additional except keeping the name as it is until they reach a non-experimental version. Please tell us if you have any furhter concerns?

> * There are different options to enable the PAM module and they have different trade-offs: https://github.com/google/fscrypt#enabling-the-pam-module-on-other-linux-distros

The currently used approach to enable `fscrypt` seems OK to me. It fits with the rest in `pam.nix`.
 
>   * I.e. we'd need to decide what's the best default for Nixpkgs.
> * I'm not sure if this belongs to `nixos/modules/security/pam.nix` (there's already a comment that "Most of these should be moved to specific modules.")

Moving to specific modules is a larger task so shouldn't we fit to the current approach until there is a volunteer to do a refactoring?

**Comment by NeoTheFox** [comment]
I also think that this is not disruptive/overreaching. Since fscrypt-experimental exists in the repos and provides the required PAM module not having a way to enable it without resorting to hacks seems incomplete to me. After all this only adds a few lines of text to the config just to enable the functionality that's already there.

**Comment by primeos** [comment]
> Could you please help us with this? I don't know who is qualified to review this.

In general I'd try to ping the users that contributed the most (or recently) to the file you're changing (listed [here](https://github.com/NixOS/nixpkgs/blob/36a3b9d6b1316bc612a639ccfa85f66519bcd0ca/nixos/modules/security/pam.nix) or via `git shortlog -sne nixos/modules/security/pam.nix`) and other active Nixpkgs reviewers (e.g. ping 1-4 potential reviewers and if there's no response withing a few days repeat that proces).
(Unfortunately the details are a bit more complicated... :o I guess we really should better document this process somewhere.)

I can try to review and help with the `fscrypt` stuff but I'm not comfortable merging this without another reviewer approving it.

> Stragely the package is not experimental in other OSes.

Yeah, the name is pretty non-standard (this is my mistake because IIRC I packaged it pretty early and at that time there where some rough edges and I didn't want anyone to accidentally loose their data without knowing the risks).

> The currently used approach to enable fscrypt seems OK to me. It fits with the rest in pam.nix.

Ok, I'll have a closer look at it.

> Moving to specific modules is a larger task so shouldn't we fit to the current approach until there is a volunteer to do a refactoring?

This PR would of course only have to "move" the fscrypt part into a specific module (if that's possible without much effort). But maybe it's also fine in `pam.nix`.
(I'll try to give this a proper review but it might have to wait until tomorrow or even the weekend. Thanks for your fast response though!)

File: nixos/modules/security/pam.nix
Comment (primeos) [inline nixos/modules/security/pam.nix:907]: Is that part actually required / what is the use-case here?

**Review by primeos — CHANGES_REQUESTED** [review]
Sorry for the delay...
I added a few style nitpicks/improvements but overall the diff LGTM (though unfortunately I also realized how unfamiliar I'm with PAM and our NixOS module for it).

**Edit:** https://github.com/google/fscrypt/releases/tag/v0.3.0 is also relevant now.

File: nixos/modules/security/pam.nix
Comment (primeos) [inline nixos/modules/security/pam.nix:907]:
```suggestion
        fscrypt.text = "auth        required    pam_unix.so";
```

Maybe this would be better (https://github.com/google/fscrypt#allowing-fscrypt-to-check-your-login-passphrase)?

File: nixos/modules/security/pam.nix
Comment (primeos) [reply nixos/modules/security/pam.nix:215]: Maybe we should also link to e.g. https://github.com/google/fscrypt/blob/master/README.md via the description?

File: nixos/modules/security/pam.nix
Comment (primeos) [reply nixos/modules/security/pam.nix:211]: Update3: It seems to me like the best approach to implement this would be `default = false` and set ` security.pam.services.$name.enableFscrypt = config.security.pam.enableFscrypt` (the actual implementation would be a bit different) for the relevant services. But I'm not sure how realistic that approach currently is for NixOS.

**Comment by lorenz** [comment]
@NeoTheFox Do you want to continue working on this? I'd like to see this merged and could take this over.

**Comment by NeoTheFox** [comment]
@lorenz feel free to take over, it's been a long time and I don't have a system to test it on anymore

**Comment by mrvik** [comment]
> I marked this as stale due to inactivity. → [More info](https://github.com/NixOS/nixpkgs/blob/master/.github/STALE-BOT.md)

Still important. Any updates on this?

**Comment by lorenz** [comment]
This has been superseded by https://github.com/NixOS/nixpkgs/pull/199587 which has since been merged and is available as of 22.11 as `security.pam.enableFscrypt`.

**Pull request closed without merging**
